A16-tcp-user-timeout.md

```diff
@@ -0,0 +1,60 @@
+TCP User Timeout
+----
+* Author(s): Yash Tibrewal
+* Approver:
+* Status: Draft
+* Implemented in:
+* Last updated: 2018-8-22
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/6VZYCFZpyTI
+
+## Abstract
+
+Provide an option to set the socket option, TCP_USER_TIMEOUT on TCP sockets for Linux kernels 2.6.37 and later.
+
+## Background
+
+TCP_USER_TIMEOUT is a socket option that has been available on Linux since kernel version 2.6.37. As described the Linux manual page, "This option takes an unsigned int as an argument. When the value is greater than 0, it specifies the maximum amount of time in milliseconds that transmitted data may remain unacknowledged before TCP will forcibly close the corresponding connection and return ETIMEDOUT to the application.  If the option value is specified as 0, TCP will to use the system default."
+
+At the time this proposal was written, gRPC Core faced an issue where data could be stuck for a very long time at the gRPC TCP layer. For cases where the network goes down, and the socket cannot accept more data for writing, an RPC could be blocked for a very long time. Even if the deadline expired on such an RPC, gRPC Core was incapable of shutting down the RPC, because the TCP layer would still have references to user data. Setting TCP_USER_TIMEOUT to a reasonable value would put a bound on the maximum time the TCP layer would wait for TCP packets to be acknowledged, and close the connection if the packets are not acknowledged in time. This would end up releasing references to user data at the gRPC TCP layer, thereby allowing the RPC to fail in a timely manner.
+
+### Related Proposals:
+[Client-side Keepalive](https://github.com/grpc/proposal/blob/master/A8-client-side-keepalive.md)
+[Server-side Connection Management](https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md)
+
+## Proposal
+
+gRPC already has a mechanism to detect connection failures, called [keepalive](https://github.com/grpc/proposal/blob/master/A8-client-side-keepalive.md). The KEEPALIVE_TIMEOUT configuration option controls the amount of time, gRPC waits for a HTTP/2 ping to be acknowledged. Since the effective aim of KEEPALIVE_TIMEOUT is similar to TCP_USER_TIMEOUT, the proposal is to reuse this configuration option to set the socket option TCP_USER_TIMEOUT. Since, TCP_USER_TIMEOUT is only available on Linux kernels 2.6.37 and later, this proposal is limited to those platforms.
+
```

Comment:
That is true, and it is also a good thing in a way since applications will not notice unwanted behavioral changes anymore due to a default TCP_USER_TIMEOUT of 20 seconds, if keepalive was disabled in the first place.

On the other hand, people who are worried about connection breakage getting detected on time would have already set keepalives, and doing it this way will allow them to continue with that.